File: ios/SiezaGym/Domain/RoutineSummary.swift
import Foundation

/// Estimacion de duracion y reparto muscular de una rutina.
/// Los segundos por serie son los mismos que usa la web, para que la app no
/// tenga dos ideas distintas de cuanto dura un entrenamiento.
nonisolated enum RoutineSummary {
    static let secondsPerWorkingSet = 40
    static let secondsPerRest = 75

    static func estimatedMinutes(_ routine: Routine, catalog: [String: Exercise]) -> Int {
        let seconds = routine.exercises.reduce(0) { total, item in
            let exercise = catalog[item.exerciseID]
            let working = exercise?.registrationType.isTimeBased == true
                ? (item.targetReps > 0 ? item.targetReps : secondsPerWorkingSet)
                : secondsPerWorkingSet
            return total + item.targetSets * (working + secondsPerRest)
        }
        return Int((Double(seconds) / 60).rounded())
    }

    struct MuscleShare: Sendable, Hashable, Identifiable {
        let muscle: MuscleGroup
        let pct: Double
        var id: MuscleGroup { muscle }
    }

    /// Los targetSets de cada ejercicio se reparten entre sus musculos segun
    /// `muscleWeights`, y el total se normaliza sobre el volumen de la rutina.
    static func muscleDistribution(_ routine: Routine, catalog: [String: Exercise]) -> [MuscleShare] {
        var raw: [MuscleGroup: Double] = [:]
        var total: Double = 0

        for item in routine.exercises {
            guard let exercise = catalog[item.exerciseID] else { continue }
            for (muscle, weight) in exercise.muscleWeights {
                let contribution = Double(item.targetSets) * weight
                raw[muscle, default: 0] += contribution
                total += contribution
            }
        }

        guard total > 0 else { return [] }
        return raw
            .map { MuscleShare(muscle: $0.key, pct: $0.value / total) }
            .sorted { $0.pct > $1.pct }
    }
}

File: android/app/src/main/kotlin/com/siezagym/app/Domain/RoutineSummary.kt
package com.siezagym.app.Domain

import com.siezagym.app.Models.Exercise
import com.siezagym.app.Models.MuscleGroup
import com.siezagym.app.Models.Routine
import kotlin.math.roundToInt

/** Estimación de duración y reparto muscular de una rutina.
 *  Los segundos por serie son los mismos que usa la web. */
object RoutineSummary {
    const val secondsPerWorkingSet = 40
    const val secondsPerRest = 75

    fun estimatedMinutes(routine: Routine, catalog: Map<String, Exercise>): Int {
        val seconds = routine.exercises.sumOf { item ->
            val exercise = catalog[item.exerciseID]
            val working = if (exercise?.registrationType?.isTimeBased == true) {
                if (item.targetReps > 0) item.targetReps else secondsPerWorkingSet
            } else secondsPerWorkingSet
            item.targetSets * (working + secondsPerRest)
        }
        return (seconds / 60.0).roundToInt()
    }

    data class MuscleShare(val muscle: MuscleGroup, val pct: Double)

    /** Los targetSets de cada ejercicio se reparten entre sus músculos según
     *  `muscleWeights`, y el total se normaliza sobre el volumen de la rutina. */
    fun muscleDistribution(routine: Routine, catalog: Map<String, Exercise>): List<MuscleShare> {
        val raw = linkedMapOf<MuscleGroup, Double>()
        var total = 0.0

        for (item in routine.exercises) {
            val exercise = catalog[item.exerciseID] ?: continue
            for ((muscle, weight) in exercise.muscleWeights) {
                val contribution = item.targetSets.toDouble() * weight
                raw[muscle] = (raw[muscle] ?: 0.0) + contribution
                total += contribution
            }
        }

        if (total <= 0) return emptyList()
        return raw.entries
            .map { MuscleShare(it.key, it.value / total) }
            .sortedByDescending { it.pct }
    }
}